File: Website/InfoTerraTerra/InfoTerraTerra_Library/Tracking/TrackingRepository.cs
using Dapper;
using InfoTerraTerra_Library.Extensions;
using InfoTerraTerra_Library.Volantini;
using Microsoft.Data.SqlClient;

namespace InfoTerraTerra_Library.Tracking;

public class TrackingRepository
{
    private readonly string _connectionString;
    private readonly VolantiniRepository _volantiniRepository;

    public TrackingRepository(
        IInfrastructureConfigurationProvider infrastructureConfigurationProvider,
        VolantiniRepository volantiniRepository)
    {
        _connectionString = infrastructureConfigurationProvider.MainSqlServerConnectionString;
        _volantiniRepository = volantiniRepository;
    }
    
    public async Task InsertQrOpenAsync(QrOpen trackingData)
    {
        await using var connection = new SqlConnection(_connectionString);
        await connection.ExecuteAsync(
            @"INSERT INTO Tracking.QrOpen 
                    (Ip, Os, Referer, Slug, IdVolantino, Citta, Via, Luogo, DateUtc)
                VALUES (@Ip, @Os, @Referer, @Slug, @IdVolantino, @Citta, @Via, @Luogo, GETUTCDATE())",
        new
        {
            trackingData.Ip,
            trackingData.Os,
            trackingData.Referer,
            trackingData.TrackingSlug.Slug,
            trackingData.TrackingSlug.IdVolantino,
            Citta = ParseUrlValue(trackingData.TrackingSlug.Citta),
            Via = ParseUrlValue(trackingData.TrackingSlug.Via),
            Luogo = ParseUrlValue(trackingData.TrackingSlug.Luogo)
        });
    }

    private static string? ParseUrlValue(string? what)
        => what
            ?.Replace("-", " ")
            ?.Replace("_", " ")
            ?.CapitalizeFirstLetter();

    public async Task<TrackingQrOpenStatistics> GetStatistics()
    {
        await using var connection = new SqlConnection(_connectionString);
        var groupedData = (await connection.QueryAsync<TrackingGroupedData>(
            @"SELECT IdVolantino, Citta, Via, Luogo, COUNT(*) AS HowMany 
                FROM Tracking.QrOpen
                WHERE IdVolantino IS NOT NULL
                GROUP BY IdVolantino, Citta, Via, Luogo
                ORDER BY IdVolantino DESC"))
            ?.ToArray();

        if (groupedData == null) 
            return new TrackingQrOpenStatistics(null);
        
        foreach (var data in groupedData)
        {
            var volantino = await _volantiniRepository.GetVolantino(data.IdVolantino);
            if (volantino != null)
            {
                data.Volantino = volantino;
            }
        }

        return new TrackingQrOpenStatistics(groupedData
            .Where(d => d.Volantino != null)
            .ToArray());
    }
}

File: dataLayer/tracking/TrackingRepository.ts
import 'server-only';
import {TrackingQrOpenStatistics} from "./TrackingQrOpenStatistics";
import StringHelper from "../../services/StringHelper";
import Configurations from "../Configurations";
import {TrackingGroupedData} from "./TrackingGroupedData";
import VolantiniRepository from "../volantini/VolantiniRepository";
import {QrOpen} from "./QrOpen";
import {Client} from "pg";

const parseUrlValue = (what: string | null | undefined): string | null => {
    return what
        ? StringHelper.capitalizeFirstLetter(what
            .replace(/-/g, " ")
            .replace(/_/g, " "))
        : null;
}

const TrackingRepository = {

    insertQrOpenAsync: async (trackingData: QrOpen) => {

        const client = new Client(Configurations.postgresConfiguration);
        await client.connect();

        try {
            await client
                .query(`
            INSERT INTO "Tracking"."QrOpen"
            ("Ip", "Os", "Referer", "Slug", "IdVolantino", "Citta", "Via", "Luogo", "DateUtc")
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, NOW())`,
                    [
                        trackingData.ip,
                        trackingData.os,
                        trackingData.referer,
                        trackingData.trackingSlug.slug,
                        trackingData.trackingSlug.idVolantino,
                        parseUrlValue(trackingData.trackingSlug.citta),
                        parseUrlValue(trackingData.trackingSlug.via),
                        parseUrlValue(trackingData.trackingSlug.luogo)
                    ]);
        }
        finally {
            await client.end();
        }
    },

    getStatisticsAsync: async (): Promise<TrackingQrOpenStatistics> => {

        const client = new Client(Configurations.postgresConfiguration);
        await client.connect();

        try {
            const result = await client
                .query(`
            SELECT "IdVolantino", "Citta", "Via", "Luogo", COUNT(*)::int AS "HowMany"
            FROM "Tracking"."QrOpen"
            WHERE "IdVolantino" IS NOT NULL
            GROUP BY "IdVolantino", "Citta", "Via", "Luogo"
            ORDER BY "IdVolantino" DESC
        `);

            const groupedData: TrackingGroupedData[] = result.rows;

            if (!groupedData || groupedData.length === 0) {
                return new TrackingQrOpenStatistics(null);
            }

            for (const data of groupedData) {
                const volantino = VolantiniRepository.getById(data.IdVolantino);

                if (volantino) {
                    data.volantino = volantino;
                }
            }

            return new TrackingQrOpenStatistics(
                groupedData.filter((d) => d.volantino !== null)
            );
        }
        finally {
            await client.end();
        }
    }
}

export default TrackingRepository;